### PR TITLE
feat: turn-on strictTemplates for Angular

### LINKED ISSUE #4744: Datepicker with reactive form and strictTemplates throws Type 'string' is not assignable to type 'Date'
## Describe the bug

The datepicker component in its reactive form implementation as shown in the [official documentation](https://clarity.design/documentation/datepicker) at the *Reactive Form Demo* section, throws the following typescript compilation error, when `strictTemplates: true` in `tsconfig.json`:

```
error TS2322: Type 'string' is not assignable to type 'Date'.
```

The app is compiled successfully when `strictTemplates` is disabled. 

## How to reproduce

1. Ensure `strictTemplates` is enabled `tsconfig.json`
```tsconfig.json
{
...
  "angularCompilerOptions": {
    ...
    "strictTemplates": true
  }
}
```
2. Use the example code, exactly as found in the [official documentation](https://clarity.design/documentation/datepicker) at the *Reactive Form Demo* section.

3. See typescript compilation error
```
ERROR in ...component.html:5:26 - error TS2322: Type 'string' is not assignable to type 'Date'.
```

## Expected behavior

The app should compile successfully as it does when `strictTemplates` is not enabled.

## Versions

**App**

- Angular: [9.1.11]
- Node: [14.4.0]
- Clarity: [3.1.4]

**Device:**

- Type: [MacBook]
- OS: [iOS > docker > node:14-alpine]
- Browser [chrome]
- Version [latest]

### PR DESCRIPTION
**Pull request by bdryanovski**
Enabled Angular compiler options for strictTemplates and resolving known errors.

When enabling the

```
"angularCompilerOptions": {
  "strictTemplates": true
}
```

Some changes were required to pass all the validation on build time.

Side-Effects (or lack of them): 
  *  No breaking changes are required to consumers of the `clr-angular` if your code was passing the `strictTemplates` before that.
  * Now by-default we will run with `strictTemplates: true` to prevent feature fixes.
  * There is a noticeable performance drop when running the build task.

Time with `strictTemplates: true` - ✨ **Done in 110.16s.**
Time with `strictTemplates: false` - ✨ Done in 73.92s.

So next time when we write an Angular Input that we expect to be used only as an argument for example:
```html
<div clrProp></div>
```
And this argument is something different than a `string` such as `number`, `boolean` etc. We have to know that `@Input` by default always returns `string` if they are not bound before that so we have to add a second type `string` and handle it inside our code.

```ts
@Input('clrProp') prop: boolean | string
```

I try to write an extra decorator to handle this for us, the issue is that we are assigning types to the `@Input` decorator and we still have to add the second type so there was no much use of just a decorator to convert any value to `boolean` or `number` 

```js
@Input('clrProp') 
@InputBoolean()
prop: boolean | string
```
The only bonus here is that `prop` will always be of type `boolean`, but to pass the compiler we still have to mention `string` as a possible value. This is not much more cleaner - so I dropped the idea.

There is an open issue for that in Angular - [it's from 2017 so I don't have big hopes for it.](https://github.com/angular/angular/issues/14761)
A proposal for solving this [issue](https://github.com/angular/angular/issues/21143) and one more [here](https://github.com/angular/angular/issues/19929)

We have an open issue for one of the errors so this PR must resolve it as well.

Fixes #4744

To test it run inside `packages/angular`
```bash
yarn build && yarn dev 
```
Everything must pass without errors;

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [ ] If applicable, have a visual design approval

## PR Type

What kind of change does this PR introduce?

- [x] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] clarity.design website / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/angular/golden/clr-angular.d.ts [hunk 1]
```diff
@@ -889,7 +889,7 @@ export declare class ClrDateInput extends WrappedFormControl<ClrDateContainer> i
     onValueChange(target: HTMLInputElement): void;
     setFocusStates(): void;
     triggerValidation(): void;
-    static ngAcceptInputType_date: Date | null;
+    static ngAcceptInputType_date: Date | null | string;
```
Comment (Jinnie) [inline packages/angular/golden/clr-angular.d.ts:892]: That's not the most supportable feature I've seen :( 
Still don't have a clue why we would need it.

**Review by Jinnie — APPROVED** [review]

File: packages/angular/projects/dev/src/app/datagrid/pagination/pagination.ts [hunk 2]
```diff
@@ -25,8 +25,8 @@ export class DatagridPaginationDemo {
     this.users = inventory.all;
   }
 
-  pageChange($event: Event) {
-    console.log($event);
+  pageChange(pageNumber: number) {
+    console.log(pageNumber);
```
Comment (bbogdanov) [inline packages/angular/projects/dev/src/app/datagrid/pagination/pagination.ts:29]: Do we need this `console.log`?

File: packages/angular/golden/clr-angular.d.ts [hunk 3]
```diff
@@ -1125,8 +1125,8 @@ export declare class ClrIfError extends AbstractIfState {
 }
 
 export declare class ClrIfExpanded implements OnInit, OnDestroy {
-    get expanded(): boolean;
-    set expanded(value: boolean);
+    get expanded(): boolean | string;
```
Comment (bbogdanov) [inline packages/angular/golden/clr-angular.d.ts:1128]: The getter is not part of an `@Input` so do we need the `string` type?

File: packages/angular/projects/dev/src/app/datagrid/pagination/pagination.ts (hunk 2 above)
Comment (bdryanovski) [reply packages/angular/projects/dev/src/app/datagrid/pagination/pagination.ts:29]: This is for demo purpose inside the devApp, so yes :)

File: packages/angular/golden/clr-angular.d.ts (hunk 3 above)
Comment (bdryanovski) [reply packages/angular/golden/clr-angular.d.ts:1128]: I don't understand the question?

Comment (bdryanovski) [reply packages/angular/golden/clr-angular.d.ts:1128]: getter and setters must return and work with the same type so they are coupled to each other.

```ts
 get expanded(): boolean | string {
    return this._expanded;
  }

  @Input('clrIfExpanded')
  set expanded(value: boolean | string) {
    if (typeof value === 'boolean') {
      this.expand.expanded = value;
      this._expanded = value;
    }
  }
```

Even that the getter method is on top of the `Input` they are linked

**Review by bbogdanov — APPROVED** [review]

**Pull request merged**